File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/features/jwks/JwksUseCase.java
package br.com.deltaglobalbank.identity.features.jwks;

import java.math.BigInteger;
import java.security.interfaces.RSAPublicKey;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

import br.com.deltaglobalbank.identity.domain.token.SigningKey;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyStatus;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.PemConverter;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class JwksUseCase {

    private final SigningKeyRepository signingKeyRepository;
    private final PemConverter pemConverter;

    public JwksUseCase(SigningKeyRepository signingKeyRepository, PemConverter pemConverter) {
        this.signingKeyRepository = signingKeyRepository;
        this.pemConverter = pemConverter;
    }

    @Transactional(readOnly = true)
    public JwkResponse execute() {
        List<SigningKey> activeKeys = signingKeyRepository.findAllByStatus(SigningKeyStatus.ACTIVE);
        List<SigningKey> retiredKeys = signingKeyRepository.findAllByStatus(SigningKeyStatus.RETIRED);

        List<Jwk> jwks = new ArrayList<>();
        activeKeys.forEach(k -> jwks.add(toJwk(k)));
        retiredKeys.forEach(k -> jwks.add(toJwk(k)));

        return new JwkResponse(jwks);
    }

    private Jwk toJwk(SigningKey signingKey) {
        RSAPublicKey publicKey = pemConverter.toPublicKey(signingKey.getPublicKey());

        return new Jwk(
            "RSA",
            "sig",
            signingKey.getAlgorithm(),
            signingKey.getKid(),
            encodeUnsignedBase64Url(publicKey.getModulus().toByteArray()),
            encodeUnsignedBase64Url(publicKey.getPublicExponent().toByteArray())
        );
    }

    private String encodeUnsignedBase64Url(byte[] bytes) {
        byte[] stripped = bytes.length > 0 && bytes[0] == 0
            ? java.util.Arrays.copyOfRange(bytes, 1, bytes.length)
            : bytes;
        return Base64.getUrlEncoder().withoutPadding().encodeToString(stripped);
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/features/jwks/JwksUseCase.kt
package br.com.deltaglobalbank.identity.features.jwks

import br.com.deltaglobalbank.identity.domain.token.SigningKey
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository
import br.com.deltaglobalbank.identity.domain.token.SigningKeyStatus
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.PemConverter
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.interfaces.RSAPublicKey
import java.util.Base64

@Service
class JwksUseCase(
    private val signingKeyRepository: SigningKeyRepository,
    private val pemConverter: PemConverter
) {

    @Transactional(readOnly = true)
    fun execute(): JwkResponse {
        val activeKeys = signingKeyRepository.findAllByStatus(SigningKeyStatus.ACTIVE)
        val retiredKeys = signingKeyRepository.findAllByStatus(SigningKeyStatus.RETIRED)

        val jwks = (activeKeys + retiredKeys).map { it.toJwk() }

        return JwkResponse(keys = jwks)
    }

    private fun SigningKey.toJwk(): Jwk {
        val publicKey = pemConverter.toPublicKey(this.publicKey) as RSAPublicKey

        return Jwk(
            kty = "RSA",
            use = "sig",
            alg = this.algorithm,
            kid = this.kid,
            n = encodeUnsignedBase64Url(publicKey.modulus.toByteArray()),
            e = encodeUnsignedBase64Url(publicKey.publicExponent.toByteArray())
        )
    }

    private fun encodeUnsignedBase64Url(bytes: ByteArray): String {
        val stripped = if (bytes.isNotEmpty() && bytes[0] == 0.toByte()) {
            bytes.copyOfRange(1, bytes.size)
        } else {
            bytes
        }
        return Base64.getUrlEncoder().withoutPadding().encodeToString(stripped)
    }
}